pkg/controller/sync/version/manager.go

```diff
@@ -0,0 +1,431 @@
+		default:
+		}
+
+		var err error
+		versionList, err = m.adapter.List(m.namespace)
+		if err != nil {
+			runtime.HandleError(fmt.Errorf("Failed to list propagated versions for %q: %v", m.templateKind, err))
+			// Do not return the error to allow the operation to be retried.
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		return nil, false
+	}
+	return versionList, true
+}
+
+func (m *VersionManager) load(versionList pkgruntime.Object, stopChan <-chan struct{}) bool {
+	typePrefix := common.PropagatedVersionPrefix(m.targetKind)
+	ok := m.adapter.Iterate(versionList, func(obj pkgruntime.Object) bool {
+		select {
+		case <-stopChan:
+			glog.V(4).Infof("Halting version manager load due to closed stop channel")
+			return false
+		default:
+		}
+
+		qualifiedName := util.NewQualifiedName(obj)
+		// Ignore propagated version for other types
+		if strings.HasPrefix(qualifiedName.Name, typePrefix) {
+			m.versions[qualifiedName.String()] = obj
+		}
+		return true
+	})
+	if !ok {
+		return false
+	}
+	m.Lock()
+	m.hasSynced = true
+	m.Unlock()
+	glog.V(4).Infof("Version manager for %q synced", m.templateKind)
+	return true
+}
+
+// versionQualifiedName derives the qualified name of a version
+// resource from the qualified name of a template or target resource.
+func (m *VersionManager) versionQualifiedName(qualifiedName util.QualifiedName) util.QualifiedName {
+	namespace := qualifiedName.Namespace
+	if m.targetKind == util.NamespaceKind {
+		namespace = qualifiedName.Name
+	}
+	versionName := common.PropagatedVersionName(m.targetKind, qualifiedName.Name)
+	return util.QualifiedName{Name: versionName, Namespace: namespace}
+}
+
+// writeVersion serializes the current state of the named propagated version to the API.
+func (m *VersionManager) writeVersion(qualifiedName util.QualifiedName) util.ReconciliationStatus {
+	key := qualifiedName.String()
+	adapterType := m.adapter.TypeName()
+
+	m.RLock()
+	obj, ok := m.versions[key]
+	m.RUnlock()
+	if !ok {
+		// Version is no longer tracked - ensure removal from API
+		err := m.deleteFromAPI(qualifiedName)
+		if err != nil {
+			runtime.HandleError(fmt.Errorf("Failed to delete %s %q: %s", adapterType, key, err))
+			return util.StatusError
+		}
+		return util.StatusAllOK
+	}
+
+	metaAccessor, err := meta.Accessor(obj)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("Failed to retrieve meta accessor for %s %q: %s", adapterType, key, err))
+		return util.StatusError
+	}
+
+	creationNeeded := len(metaAccessor.GetResourceVersion()) == 0
+	if creationNeeded {
+		createdObj, err := m.adapter.Create(obj)
+		if errors.IsAlreadyExists(err) {
+			// Version was written to the API after the version manager loaded
+			glog.V(4).Infof("Refreshing %s %q from the API due to already existing", adapterType, key)
+			err := m.refreshVersion(obj)
+			if err != nil {
+				runtime.HandleError(fmt.Errorf("Failed to refresh existing %s %q from the API: %v", adapterType, key, err))
```

Comment:
If the version in the cache and storage is deleted after the `Create` on line 287, but before we call `refreshVersion` on line 291, we would inadvertently throw this error.

Reply:
For `NotFound` to be returned from `refresh`, the resource would have to have been created before `Create` and then deleted before `Get` was called in `refreshVersion`.  I think the probability of this happening is low enough that it's not worth adding special-case handling.